docs/source_rewriter.md

```diff
@@ -0,0 +1,210 @@
+
+First a header's AST can change depending on the preprocessor context when
+it's included. We currently create one rewritten `.h` for each input `.h`
+which can have correctness issues in some rare cases. While we can fix the
+correctness issue by creating one rewritten `.h` for each `#include` this
+engineering problem will quickly become messy. We also use the same set of `-D`
+for all `.h`s, but the correct solution is to use the exact set of flags used to
+compile the `.c` which includes each `.h`.
+
+The second problem is that code that makes heavy use of function pointers is
+burdensome to annotate and they degrade readability. Even in the case where only a
+small set of the function pointers in a code base are used, the rewritten `.h`s
+trigger compiler errors at every site without annotations. Also some of the
+annotations need the pointer's function signature specified with the mangled
+type name which is a non-starter.
+
+## Goals
+
+By analyzing and rewriting `.c`s we can avoid these issues. The goal here will
+be to (1) create an output `.c` for call gate wrappers and (2) insert simplified
+function pointer annotations in the input `.c`s. We'll initially focus on the
+case with one rewritten `.c` for each input `.c` and attempt to insert the
+annotations such that the rewritten `.c` can be compiled with the same flags.
+
+## Source rewriting process
+
+The goals above can be broken down into these subgoals.
+
+1. Define call gates for direct compartment transitions
+2. Add call gates for pointers
+3. Add call gates for indirect callsites
+
+The rewriter will take all source files as input along with which compartment
+each belongs to. It will generate one output `.c` and one `.h` for use with all
+compartments. It will also generate compartment-specific linker args and
+`objcopy --redefine-syms` files.
+
+### Define call gates for direct transitions
+
+The rewriter will work in two phases. In the first phase it'll parse all source
+files and keep track of functions defined and their compartment, functions
+called and from which compartment (TODO: Should we check functions called or
+undefined?). Calls to functions not defined in any of the
+input sources (e.g. libc) will be assumed to remain in the caller's
+compartment (TODO: What if a user forgets an input .c? maybe assert that these
+functions are declared in /usr/ headers).
+
+In phase two we'll generate the call gates for each direct transition. Since we
+know all the caller and target compartment we can hard-code the pkey in each
+call gate wrapper. These wrappers are inserted with `ld --wrap` so their names
+must start with `__wrap_`. Since a function may be called from multiple
+compartments and call gates have a hardcoded caller pkey, we need one wrapper
+per caller compartment. To distinguish wrappers for different compartments we'll
+name the wrappers `__wrap_$FUNCTION_NAME_from_$CALLER_PKEY`. This will require
```

Comment:
What happened to defining these symbols with hidden visibility? I.e. define `__wrap_foo` separately in compartment1 and compartment2, each with hidden visibility then linking that into the DSO?